### DISCUSSION TITLE
First form submit with errors returns an empty error object.

### DISCUSSION THREAD

**Discussion by Zcehtro** [react-hook-form/react-hook-form#9396] [Q&A] [answered]
I'm testing how my form handles errors but I'm having terrible luck so far. Every time I hit the submit button the first time, the `errors` object returns empty `{}`. If I hit the submit button a second time and more times then on, the `errors` object returns the correct content and keys.

But every time I reload the app, the `errors` object returns empty again every first time I hit submit.

This is my React Native code sans the imports.

```typescript
export const ContactUs = () => {
  const navigation = useNavigation<StackNavigation>();

  const {
    control,
    handleSubmit,
    formState: { errors },
    reset,
  } = useForm<IFormContactUs>({
    defaultValues: {
      name: '',
      surname: '',
      phone: '',
      email: '',
      message: '',
    },
    mode: 'onSubmit',
    reValidateMode: 'onSubmit',
  });

  const onSubmit: SubmitHandler<IFormContactUs> = async (data) => {
    const res = await submitContactUs(data);

    if (res.request.status >= 200 && res.request.status <= 399) {
      Alert('Your request has been sent.');
      reset();
      navigation.navigate('RegisterLogin');
    } else {
      Alert('Error. Please try again.');
    }
  };

  const onError = async () => {
   const errorMsgs = Object.keys(errors)
      .reduce(
        (a: string, field: string) => (errors[field] ? a + '\n' + errors[field].message : a),
        ''
      )
      .trim();
   Alert('Error. Please fix the following errors:\n\n' + errorMsgs);
  };
```

According to my code above, the Alert box should print all the errors, but it only does so after the second submit:

![photo_2022-11-18_17-26-10](https://user-images.githubusercontent.com/10564090/202796689-c1043b09-62aa-4ba1-8fa7-48393711bcee.jpg)

![photo_2022-11-18_17-29-41](https://user-images.githubusercontent.com/10564090/202796712-18f9dc40-a6a7-4240-8e84-0af97be01fb5.jpg)

Why is this happening? How can I get the `errors` object to print correctly on the first try?

**Comment by bluebill1049** [comment]
Could you create an expo snack on this?

**Reply by Zcehtro** [reply]
Sure! Give me a few minutes...

**Reply by Zcehtro** [reply]
Apologies, it was easier for me to simply create a repo.

Here's the link: https://github.com/Zcehtro/contact-us-demo-snack

**Comment by leapful** [comment, marked as answer]
@Zcehtro

The callback `onInvalid` of method `handleSubmit` from react-hook-form is triggered right after there is an error while validating form values but `formState.errors` is only being updated on the next re-rendering as it belongs to hook `useForm`.

You should be first parameter of `onInvalid(errors)` instead of `formState.errors`.

**Reply by Zcehtro** [reply]
You're absolutely right! I figured this out the following day after posting this but work made me forget completely.

I ended up ditching the `errors` state and just used the `onError` callback from `handleSubmit`. The form now works as I intended. Thx anyway for the reply!

**Comment by Phibee** [comment]
@Zcehtro, your code is all goods bro. just update your onError callback. coz it has 2 arguments; 
(errors: Object, e?: Event)
or you may refer to this documentation
https://react-hook-form.com/api/useform/handlesubmit/

![ezgif-5-41441f0d86](https://user-images.githubusercontent.com/3889714/204565720-5017bc8f-eabd-4d08-bd71-9ce34d76b2ef.gif)

**Reply by Zcehtro** [reply]
Tvm for the reply mate! You're right, I had to use the `onError` function, but I wasn't aware that I could set it up as a callback for `handleSubmit`. So I did precisely that!
